### PR TITLE
Partial revert of #8904 for MaskedArray and numpy <= 1.15

### LINKED ISSUE #9388: Astropy 3.2.2 test failures with Numpy 1.14.6
Some of the tests for Astropy 3.2.2 are failing with Numpy 1.14.6 which is causing the wheels to not be built for Python 3.7 (see https://github.com/astropy/astropy/issues/9371 for example). To reproduce with Python 3.7:

```
python -m venv python/debug-wheel
source ~/python/debug-wheel/bin/activate
pip install numpy==1.14.6
pip install astropy --no-binary :all:
pip install pytest-astropy
python -c "import astropy;astropy.test(args='-k test_regression_futuretimes_4302')"```
```

The error is then:

```
__________________________________________________________ test_regression_futuretimes_4302 ___________________________________________________________

    def test_regression_futuretimes_4302():
        """
        Checks that an error is not raised for future times not covered by IERS
        tables (at least in a simple transform like CIRS->ITRS that simply requires
        the UTC<->UT1 conversion).
    
        Relevant comment: https://github.com/astropy/astropy/pull/4302#discussion_r44836531
        """
        from astropy.utils.exceptions import AstropyWarning
    
        # this is an ugly hack to get the warning to show up even if it has already
        # appeared
        from astropy.coordinates.builtin_frames import utils
        if hasattr(utils, '__warningregistry__'):
            utils.__warningregistry__.clear()
    
        with catch_warnings() as found_warnings:
            future_time = Time('2511-5-1')
            c = CIRS(1*u.deg, 2*u.deg, obstime=future_time)
>           c.transform_to(ITRS(obstime=future_time))

python/debug-wheel/lib/python3.7/site-packages/astropy/coordinates/tests/test_regression.py:226: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
python/debug-wheel/lib/python3.7/site-packages/astropy/coordinates/baseframe.py:1202: in transform_to
    return trans(self, new_frame)
python/debug-wheel/lib/python3.7/site-packages/astropy/coordinates/transformations.py:1377: in __call__
    curr_coord = t(curr_coord, curr_toframe)
python/debug-wheel/lib/python3.7/site-packages/astropy/coordinates/transformations.py:980: in __call__
    return supcall(fromcoord, toframe)
python/debug-wheel/lib/python3.7/site-packages/astropy/coordinates/builtin_frames/intermediate_rotation_transforms.py:79: in cirs_to_itrs
    pmat = cirs_to_itrs_mat(itrs_frame.obstime)
python/debug-wheel/lib/python3.7/site-packages/astropy/coordinates/builtin_frames/intermediate_rotation_transforms.py:31: in cirs_to_itrs_mat
    xp, yp = get_polar_motion(time)
python/debug-wheel/lib/python3.7/site-packages/astropy/coordinates/builtin_frames/utils.py:41: in get_polar_motion
    xp, yp, status = iers.IERS_Auto.open().pm_xy(time, return_status=True)
python/debug-wheel/lib/python3.7/site-packages/astropy/utils/iers/iers.py:660: in open
    cls.iers_table = cls.read(file=filename)
python/debug-wheel/lib/python3.7/site-packages/astropy/utils/iers/iers.py:507: in read
    table = cls._combine_a_b_columns(iers_a)
python/debug-wheel/lib/python3.7/site-packages/astropy/utils/iers/iers.py:469: in _combine_a_b_columns
    is_predictive = (table['UT1Flag_A'] == 'P') | (table['PolPMFlag_A'] == 'P')
python/debug-wheel/lib/python3.7/site-packages/astropy/table/column.py:986: in _compare
    return result.data if isinstance(result, Column) else result
python/debug-wheel/lib/python3.7/site-packages/astropy/table/column.py:1278: in data
    out.fill_value = self.fill_value
python/debug-wheel/lib/python3.7/site-packages/numpy/ma/core.py:3375: in __setattr__
    super(MaskedArray, self).__setattr__(attr, value)
python/debug-wheel/lib/python3.7/site-packages/numpy/ma/core.py:3659: in set_fill_value
    target = _check_fill_value(value, self.dtype)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

fill_value = 'N', ndtype = dtype('bool')

    def _check_fill_value(fill_value, ndtype):
        """
        Private function validating the given `fill_value` for the given dtype.
    
        If fill_value is None, it is set to the default corresponding to the dtype.
    
        If fill_value is not None, its value is forced to the given dtype.
    
        The result is always a 0d array.
        """
        ndtype = np.dtype(ndtype)
        fields = ndtype.fields
        if fill_value is None:
            fill_value = default_fill_value(ndtype)
        elif fields:
            fdtype = [(_[0], _[1]) for _ in ndtype.descr]
            if isinstance(fill_value, (ndarray, np.void)):
                try:
                    fill_value = np.array(fill_value, copy=False, dtype=fdtype)
                except ValueError:
                    err_msg = "Unable to transform %s to dtype %s"
                    raise ValueError(err_msg % (fill_value, fdtype))
            else:
                fill_value = np.asarray(fill_value, dtype=object)
                fill_value = np.array(_recursive_set_fill_value(fill_value, ndtype),
                                      dtype=ndtype)
        else:
            if isinstance(fill_value, basestring) and (ndtype.char not in 'OSVU'):
                err_msg = "Cannot set fill value of string with array of dtype %s"
>               raise TypeError(err_msg % ndtype)
E               TypeError: Cannot set fill value of string with array of dtype bool

python/debug-wheel/lib/python3.7/site-packages/numpy/ma/core.py:473: TypeError
```

I have bisected this and it looks like this comes from https://github.com/astropy/astropy/pull/8904.

@mhvk, since this was your PR, is there any chance you could look into this? I know we aren't supporting Numpy 1.14 anymore in master, but it would be good to get all tests passing and do a last 3.2.x release before 4.0.

### PR DESCRIPTION
**Pull request by mhvk**
fixes #9388

@astrofrog - Note that I didn't add a changelog entry (as somehow it was at 3.2.2). I may also be targeting the wrong branch - feel free to adjust (or let me know what to do).

### CODE REVIEW

**Comment by astrofrog** [comment]
@mhvk - I saw issues with 1.14.5 as well so I'm not sure if we should be excluding <1.14.6 from the fix?

**Comment by mhvk** [comment]
OK, had to rebase anyway as my quickly-before-I-have-to-be-off strategy failed. Now 1.14.5<=numpy<1.15 gets the fix.

**Comment by bsipocz** [comment]
this doesn't effect lts, right?

**Comment by mhvk** [comment]
@bsipocz - no, no issue for lts.

**Comment by astrofrog** [comment]
The bug is also with Numpy 1.14.4 - I didn't mean to imply before that it appeared in 1.14.5 or 6 - in fact it could be in earlier versions too. So I don't think we can easily just use the first version that works since that might rule out all of 1.14.x (which is declared as the minimum compatible version). So I do think we should fix this in the branch.

**Comment by bsipocz** [comment]
OK. Then I suggest we merge this, and will see whether there'll another release or not.

File: astropy/table/column.py [hunk 1]
```diff
@@ -980,6 +981,12 @@ def _compare(self, other):
             if self.dtype.char == 'S':
                 other = self._encode_str(other)
 
+            if NUMPY_LT_1_15 and not NUMPY_LT_1_14_5 and (
```
Comment (astrofrog) [inline astropy/table/column.py]: I think we should get rid of the ``not NUMPY_LT_1_14_5`` here as the issue is present in other versions, e.g. 1.14.4

File: astropy/table/column.py [hunk 2]
```diff
@@ -980,6 +981,12 @@ def _compare(self, other):
...
+                    isinstance(self, np.ma.MaskedArray) or
+                    isinstance(other, np.ma.MaskedArray)):
+                # On numpy 1.14.6, MaskedArray subclass comparison is broken.
```
Comment (astrofrog) [inline astropy/table/column.py]: Comment needs updating

**Comment by mhvk** [comment]
@astrofrog - OK, make both changes.

**Comment by bsipocz** [comment]
The failure might be related, or something we already have on the branch (that I don't think so as the latest build for the release was clean). I'll investige when/if we go ahead with a release while keeping this open.

**Comment by bsipocz** [comment]
OK, so let's get this merged,  I'll fix up stuff if needed on the branch.

**Comment by bsipocz** [comment]
Thanks @mhvk!

**Comment by bsipocz** [comment]
I have a suspicion that the current failures I see on the wheels for v3.2.3, using numpy 1.13.3 is caused by this:

https://travis-ci.org/MacPython/astropy-wheels/jobs/603420042#L661

```
=================================== FAILURES ===================================

______________ TestColumn.test_quantity_comparison[MaskedColumn] _______________

self = <astropy.table.tests.test_column.TestColumn object at 0x7fca317c9908>

Column = <class 'astropy.table.column.MaskedColumn'>

    def test_quantity_comparison(self, Column):

        # regression test for gh-6532

        c = Column([1, 2100, 3], unit='Hz')

        q = 2 * u.kHz

>       check = c < q

/venv/lib/python3.5/site-packages/astropy/table/tests/test_column.py:159: 

_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

/venv/lib/python3.5/site-packages/astropy/table/column.py:987: in _compare

    return getattr(self.data, op)(other)

/venv/lib/python3.5/site-packages/astropy/units/quantity.py:444: in __array_ufunc__

    converters, unit = converters_and_unit(function, method, *inputs)

_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

function = <ufunc 'less'>, method = '__call__'

args = (masked_BaseColumn(data = [1 2100 3],

                  mask = [False False False],

            fill_value = 999999)

, <Quantity 2.0 kHz>)

ufunc_helper = <function helper_twoarg_comparison at 0x7fca2c162488>

units = [None, Unit("kHz")], converters = [False, None]

    def converters_and_unit(function, method, *args):

        """Determine the required converters and the unit of the ufunc result.

    

        Converters are functions required to convert to a ufunc's expected unit,

        e.g., radian for np.sin; or to ensure units of two inputs are consistent,

        e.g., for np.add.  In these examples, the unit of the result would be

        dimensionless_unscaled for np.sin, and the same consistent unit for np.add.

    

        Parameters

        ----------

        function : `~numpy.ufunc`

            Numpy universal function

        method : str

            Method with which the function is evaluated, e.g.,

            '__call__', 'reduce', etc.

        *args : Quantity or other ndarray subclass

            Input arguments to the function

    

        Raises

        ------

        TypeError : when the specified function cannot be used with Quantities

            (e.g., np.logical_or), or when the routine does not know how to handle

            the specified function (in which case an issue should be raised on

            https://github.com/astropy/astropy).

        UnitTypeError : when the conversion to the required (or consistent) units

            is not possible.

        """

    

        # Check whether we support this ufunc, by getting the helper function

        # (defined in helpers) which returns a list of function(s) that convert the

        # input(s) to the unit required for the ufunc, as well as the unit the

        # result will have (a tuple of units if there are multiple outputs).

        ufunc_helper = UFUNC_HELPERS[function]

    

        if method == '__call__' or (method == 'outer' and function.nin == 2):

            # Find out the units of the arguments passed to the ufunc; usually,

            # at least one is a quantity, but for two-argument ufuncs, the second

            # could also be a Numpy array, etc.  These are given unit=None.

            units = [getattr(arg, 'unit', None) for arg in args]

    

            # Determine possible conversion functions, and the result unit.

            converters, result_unit = ufunc_helper(function, *units)

    

            if any(converter is False for converter in converters):

                # for multi-argument ufuncs with a quantity and a non-quantity,

                # the quantity normally needs to be dimensionless, *except*

                # if the non-quantity can have arbitrary unit, i.e., when it

                # is all zero, infinity or NaN.  In that case, the non-quantity

                # can just have the unit of the quantity

                # (this allows, e.g., `q > 0.` independent of unit)

                try:

                    # Don't fold this loop in the test above: this rare case

                    # should not make the common case slower.

                    for i, converter in enumerate(converters):

                        if converter is not False:

                            continue

                        if can_have_arbitrary_unit(args[i]):

                            converters[i] = None

                        else:

                            raise UnitConversionError(

                                "Can only apply '{0}' function to "

                                "dimensionless quantities when other "

                                "argument is not a quantity (unless the "

                                "latter is all zero/infinity/nan)"

>                               .format(function.__name__))

E                               astropy.units.core.UnitConversionError: Can only apply 'less' function to dimensionless quantities when other argument is not a quantity (unless the latter is all zero/infinity/nan)

/venv/lib/python3.5/site-packages/astropy/units/quantity_helper/converters.py:189: UnitConversionError
```

**Comment by bsipocz** [comment]
python 3.7 wheels are fine, as well as all the windows ones.

**Comment by bsipocz** [comment]
Yes, I'm afraid removing this patch from the branch, make the wheel production happy.

I'm very much inclined thus to release 3.2.3 without this. However, now I'm going home, so will only continue with the release tomorrow late morning, so if you have any good and quick plans to proceed, please let me know (e.g. we need to release because of the IERS urls).

**Comment by astrofrog** [comment]
Just to be clear this is also the same failure we are seeing with the CI in astropy for this PR:

https://travis-ci.org/astropy/astropy/builds/602090265?utm_medium=notification&utm_source=github_status

So if the CI here can be fixed it would also fix the wheels.

**Comment by mhvk** [comment]
I think this is my mistake: I reverted the solution for the bug that quantity-column comparisons were not correct, but didn't skip the test that checked that comparison when one has NUMPY_LT_1_15. So, I think all it needs is a `pytest.xfail(NUMPY_LT_1_15, reason="on older numpy, we cannot solve this.")`

**Comment by mhvk** [comment]
Reverting and not solving #9388 (perhaps just testing on numpy 1.15) might be the easy way out. Do we really care with 4.0 nearly ready?

**Comment by bsipocz** [comment]
> Do we really care with 4.0 nearly ready?

yes, we need to get a new releases out today, thanks to the IERS servers. Otherwise I would agree with you totally, that this is only a teeny tiny issue, and we better focus on 4.0.

**Comment by mhvk** [comment]
OK, thanks @bsipocz and @astrofrog for all the hard, extra work!

**Pull request merged**